### ISSUE TITLE
JSON encoder for CollectD events

### ISSUE THREAD

**Issue by panda87** [mre/kafka-influxdb#6]
Hi @mre 

Please find the below several JSON examples of CollectD events.
If you have estimation\direction when\whether you build this encoder, please share, this will be very helpful feature.

```
[{"values":[1.13385872826296],"dstypes":["gauge"],"dsnames":["value"],"time":1444745136.182,"interval":10.000,"host":"xx.example.internal","plugin":"memory","plugin_instance":"","type":"percent","type_instance":"slab_recl"}]
[{"values":[0.907258064516129],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"0","type":"percent","type_instance":"user"}]
[{"values":[0.705645161290323],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"0","type":"percent","type_instance":"system"}]
[{"values":[0],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"0","type":"percent","type_instance":"wait"}]
[{"values":[37.7016129032258],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"0","type":"percent","type_instance":"nice"}]
[{"values":[0],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"0","type":"percent","type_instance":"interrupt"}]
[{"values":[0.30241935483871],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"0","type":"percent","type_instance":"softirq"}]
[{"values":[0],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"0","type":"percent","type_instance":"steal"}]
[{"values":[60.383064516129],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"0","type":"percent","type_instance":"idle"}]
[{"values":[0.709939148073022],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"1","type":"percent","type_instance":"user"}]
[{"values":[0.608519269776876],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"1","type":"percent","type_instance":"system"}]
[{"values":[0],"dstypes":["gauge"],"dsnames":["value"],"time":1444745144.824,"interval":10.000,"host":"xx.example.internal","plugin":"cpu","plugin_instance":"1","type":"percent","type_instance":"wait"}]
```

Thanks
D.

**Comment by DefunctExodus** [comment]
This would be really nice to have.

**Comment by mre** [comment]
@panda87 I've integrated an initial version into master.
Used your example messages as unit tests. See https://github.com/mre/kafka-influxdb/blob/master/kafka_influxdb/tests/encoder_test/test_collectd_json_encoder.py

The encoder uses ultrajson on Python 2 and Python 3.
This gives a nice performance boost over the standard json module.
For PyPy we fall back to the standard json module because ultrajson does not work there.
See https://github.com/esnme/ultrajson/issues/98 for more info.

I would be very glad if you could test this and give some feedback.
Don't expect much yet, as I still need to work on sending the retention period and additional fields.

**Comment by mre** [comment]
@panda87, @DefunctExodus could one of you guys test it to see if it works as expected? :-)

**Comment by DefunctExodus** [comment]
Was just talking about this today, will give it a shot on Thursday when I
am back in the office.

**Comment by mre** [comment]
@DefunctExodus awesome. Thanks! :v:

**Comment by DefunctExodus** [comment]
@mre 

Sorry ended up getting backlogged at work, Looking at this right now and most of our counters are RX / TX values so for example
``` json 
{"values":[0,0],"dstypes":["derive","derive"],"dsnames":["rx","tx"],"time":1448385766.629,"interval":30.000,"host":"hostNameHere","plugin":"snmp","plugin_instance":"","type":"if_packets","type_instance":"Stuff"}] ```


Brief look at the encoder and it appears based on your sample payload it is currently only setup to support single values.

**Comment by mre** [comment]
This pull request supports multiple values: https://github.com/mre/kafka-influxdb/pull/29
Haven't tried it though. If you want to have a look, that would be a great help.

**Comment by DefunctExodus** [comment]
Sure, I will let you know next week. Have about 80gigs of multivalue data
to parse

Thanks

**Comment by DefunctExodus** [comment]
So I am going through this and a collectd event such as this 
```python
format_json: value_list_to_json: buffer = ,{
	"values":[0,0],
	"dstypes":["derive","derive"],
	"dsnames"["rx","tx"],
	"time":1453143730.697,
	"interval":30.000,
	"host":"somehost",
	"plugin":"snmp",
	"plugin_instance":"",
	"type":"if_dropped",
	"type_instance":"someinterface"}
```

returns in an influxdb record such as this

```
2016-01-18T18:08:40Z	"somehost"	"someinterface"	"0 0"
```

I am not sure if that is the expected result or not honestly, with the values combined into a single row

**Comment by mre** [comment]
Okay, thanks for testing. Could you show me how you queried the data? I want to try it myself as well.

**Comment by DefunctExodus** [comment]
Was a simple select * on the series

Okay, thanks for testing. Could you show me how you queried the data? I
want to try it myself as well.

—
Reply to this email directly or view it on GitHub
<https://github.com/mre/kafka-influxdb/issues/6#issuecomment-172662269>.

**Comment by mre** [comment]
Although this issue is quite old by now, the bug should be fixed in master and in version 0.7.3.
Thanks to @kaffeiflex for working on this. 
The new version can be installed with `pip install kafka_influxdb` or by cloning the master repo. 

You can give it another shot, but I could also understand if you already found a workaround or another solution.
